src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgery.cs

```diff
@@ -325,35 +325,45 @@ private bool IsValidationRequired(HttpContext httpContext)
             return true;
         }
 
-        private AntiforgeryContext GetCookieTokens(HttpContext httpContext)
+        private static IAntiforgeryFeature GetAntiforgeryFeature(HttpContext httpContext)
         {
-            var services = httpContext.RequestServices;
-            var contextAccessor = services.GetRequiredService<IAntiforgeryContextAccessor>();
-            if (contextAccessor.Value == null)
+            var antiforgeryFeature = httpContext.Features.Get<IAntiforgeryFeature>();
+            if (antiforgeryFeature == null)
             {
-                contextAccessor.Value = new AntiforgeryContext();
+                antiforgeryFeature = new AntiforgeryFeature();
+                httpContext.Features.Set(antiforgeryFeature);
             }
 
-            var antiforgeryContext = contextAccessor.Value;
-            if (antiforgeryContext.HaveGeneratedNewCookieToken)
+            return antiforgeryFeature;
+        }
+
+        private IAntiforgeryFeature GetCookieTokens(HttpContext httpContext)
+        {
+            var antiforgeryFeature = GetAntiforgeryFeature(httpContext);
+            if (antiforgeryFeature == null)
```

Comment:
Yeah, and why would we want to `new` up an `AntiforgeryFeature` and never store it in the `HttpContext`?